cirq/sim/sparse_simulator.py

```diff
@@ -134,34 +135,39 @@ def _run(self, circuit: circuits.Circuit,
         param_resolver = param_resolver or study.ParamResolver({})
         resolved_circuit = protocols.resolve_parameters(circuit, param_resolver)
         self._check_all_resolved(resolved_circuit)
-
-        def measure_or_mixture(op):
-            return not protocols.has_unitary(op) and (
-                protocols.is_measurement(op) or protocols.has_channel(op))
-
-        if resolved_circuit.are_all_matches_terminal(measure_or_mixture):
-            return self._run_sweep_terminal_sample(resolved_circuit,
-                                                   repetitions)
-        return self._run_sweep_repeat(resolved_circuit, repetitions)
-
-    def _run_sweep_terminal_sample(self, circuit: circuits.Circuit,
-                                   repetitions: int) -> Dict[str, np.ndarray]:
-        for step_result in self._base_iterator(
-                circuit=circuit,
-                qubit_order=ops.QubitOrder.DEFAULT,
-                initial_state=0,
-                perform_measurements=False):
+        qubit_order = sorted(resolved_circuit.all_qubits())
+
```

Comment:
just to make sure I understand: the bug was that intermediate channel ops that could be monte carlo sampled were allowed to not be terminal, and the new way this works it splits into the biggest possible unitary chunk and if then "others" to figure out whether this is possible.

Reply:
Yes, that was the issue.